File: ruby-call-signature/src/main/java/org/jetbrains/ruby/codeInsight/types/signature/ClassInfo.java
package org.jetbrains.ruby.codeInsight.types.signature;

import org.jetbrains.annotations.NotNull;

public final class ClassInfo {
    @NotNull
    private final String myClassFQN;

    public ClassInfo(@NotNull String classFQN) {
        myClassFQN = classFQN;
    }

    @NotNull
    public String getClassFQN() {
        return myClassFQN;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        final ClassInfo classInfo = (ClassInfo) o;

        return myClassFQN.equals(classInfo.myClassFQN);
    }

    @Override
    public int hashCode() {
        return myClassFQN.hashCode();
    }
}

File: ruby-call-signature/src/main/java/org/jetbrains/ruby/codeInsight/types/signature/ClassInfo.kt
package org.jetbrains.ruby.codeInsight.types.signature


interface ClassInfo {
    val gemInfo: GemInfo?
    val classFQN: String
    

    data class Impl(override val gemInfo: GemInfo?, override val classFQN: String) : ClassInfo

    fun validate(): Boolean {
        if (classFQN.length > LENGTH_OF_FQN) {
            return false
        }
        val gemInfoVal = gemInfo
        return gemInfoVal == null || gemInfoVal.validate()
    }

    companion object {
        val LENGTH_OF_FQN = 200
    }

}


fun ClassInfo(gemInfo: GemInfo?, classFQN: String) = ClassInfo.Impl(gemInfo, classFQN)

fun ClassInfo(classFQN: String) = ClassInfo.Impl(null, classFQN)

fun ClassInfo(copy: ClassInfo) = with(copy) { ClassInfo.Impl(gemInfo?.let { GemInfo(it) }, classFQN) }

